### PR TITLE
Add `Exemplar` to metricdata package

### PR DESCRIPTION
**Pull request by MrAlias**
Part of #3834 

- Add the new `Exemplar[N int64 | float64]` `struct` to `sdk/metric/metricdata`
- Add `Exemplars []Exemplar[N]` field to `DataPoint` in `sdk/metric/metricdata`
- Add `Exemplars []Exemplar[N]` field to `HistogramDataPoint` in `sdk/metric/metricdata`
  - Update `HistogramDataPoint` to be defined generically as `HistogramDataPoint[N int64 | float64]`
  - Update `Histogram` to be defined generically as `Histogram[N int64 | float64]`
- Update `metricdatatest` to support `Exemplar`
- Update `metricdatatest` to support `HistogramDataPoint[N]`
- Update `sdk/metric` histogram aggregator to produce `Histogram[N]`
- Update exporters (stdout, Prometheus, OTLP) to compile with `Histogram[N]`
  - These exporters still need to be updated to export `Exemplar`
- Update OpenCensus bridge to support `Histogram[N]` (only produced `Histogram[float64]`)

### CODE REVIEW

File: sdk/metric/metricdata/data.go [hunk 1]
```diff
@@ -126,6 +132,9 @@ type HistogramDataPoint struct {
 	Max Extrema
 	// Sum is the sum of the values recorded.
 	Sum float64
+
+	// Exemplars is the sampled Exemplars collected during the timeseries.
+	Exemplars []Exemplar[N] `json:",omitempty"`
```
Comment (pellared) [inline sdk/metric/metricdata/data.go:137]: I think we should also change the types of `Bounds`, `Min`, `Max`, Sum` to respect the type parameter. Separate PR?

**Review by pellared — COMMENTED** [review]
General question, just to clarify: do we want to implement "Exemplar"before it is Stable in the specification. 

Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar

File: sdk/metric/metricdata/data.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/metricdata/data.go:137]: Yeah, agreed, and yeah also separate PR. :+1:

**Comment by MrAlias** [comment]
> General question, just to clarify: do we want to implement "Exemplar" before it is Stable in the specification?
> 
> Reference: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exemplar

I'm not sure we want to implement it given the unstable nature of the specification around it. That and I think the specification has some evolving to do regarding the exemplar.

That said, I do want to make sure we can add support in the future. These changes are necessary now so we don't have to implement something sub-par to support them in the future or introduce backwards incompatible changes later.

As pointed out in your other comment, I think there is still some follow-up work to address here.

1. The other fields of the `HistogramDataPoint` can now utilize the generic argument to give native values
2. The exporters can be updated to "support" exemplars as they will be in the data structure passed to them (though the SDK will not produce them yet)

I think (2) is helpful on the path to supporting exemplars, but not needed for the GA of the specification.

File: sdk/metric/metricdata/metricdatatest/assertion.go [hunk 2]
```diff
@@ -66,6 +71,14 @@ func IgnoreTimestamp() Option {
 	})
 }
 
+// IgnoreExemplars disables checking if Exemplars are different.
+func IgnoreExemplars() Option {
```
Comment (pellared) [inline sdk/metric/metricdata/metricdatatest/assertion.go:75]: There is no unit test for this option.

**Review by MadVikingGod — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Pull request merged**